\documentclass{article}
\usepackage[hidelinks]{hyperref}
\usepackage[nonumberlist]{glossaries}
\usepackage{apacite}
\let\cite\shortcite
\usepackage{authblk}
\usepackage{graphicx}
\usepackage{a4wide}
\setlength\parindent{0pt}

\bibliographystyle{apacite}
\makeglossaries
\newacronym{PD}{PD}{Parkinson's Disease}

% For drafts
\usepackage{setspace}
\doublespacing
\usepackage{lineno}
\linenumbers

\title{TITLE}
\author[1]{Theo Portlock}
%\author[1,2]{ }
\affil[1]{Science for Life Laboratory, Royal Institute of Technology (KTH), Stockholm, Sweden.}
%\affil[2]{Centre for Host Microbiome Interactions, Faculty of Dentistry, Oral \& Craniofacial Sciences, Kingâ€™s College London, London, UK.}

%\author{\vspace{-5ex}}
\date{\vspace{-5ex}}
\begin{document}
\maketitle
\tableofcontents
\printglossaries

\section{Abstract}

\section{Background}
\gls{PD}

\section{Results}
%\begin{figure}
%\centering
%\includegraphics[scale=0.8]{figures/composition.pdf}
%\caption[Compositional changes]{
	%Compositional changes. Species change with the disease}
%\label{Fcomposition}
%\end{figure}
%\autoref{Fcomposition}

\section{Discussion}

\section{Conclusion}

\section{Methods}

\subsection{Metagenomic sequence data}

\subsection{Gene abundance profiling}

\subsection{MSPs taxonomic annotation}

\subsection{MSPs functional annotation}

\subsection{Correlation analysis}

\section{Data availability}

\section{Code availability}

\section{Author Contributions}

\section{Acknowledgments}

\bibliography{library}

\section{Supplimentary material}

\end{document}
